### ISSUE TITLE
fuzz-replacer: Timeout in fuzz-replacer

### ISSUE THREAD

**Issue by mohammed90** [caddyserver/caddy#4170]
Detailed Report: https://oss-fuzz.com/testcase?key=6072821525643264

Project: caddy
Fuzzing Engine: libFuzzer
Fuzz Target: fuzz-replacer
Job Type: libfuzzer_asan_caddy
Platform Id: linux

Crash Type: Timeout (exceeds 60 secs)
Crash Address: 
Crash State:
  fuzz-replacer
  
Sanitizer: address (ASAN)

Crash Revision: https://oss-fuzz.com/revisions?job=libfuzzer_asan_caddy&revision=202105130617

Reproducer Testcase: [clusterfuzz-testcase-minimized-fuzz-replacer-6072821525643264](https://github.com/caddyserver/caddy/files/6488340/clusterfuzz-testcase-minimized-fuzz-replacer-6072821525643264.txt)

Issue on oss-fuzz tracker: [Issue 34275](https://crbug.com/oss-fuzz/34275)

**Comment by Mohammed90** [comment]
Sharing some findings... I found those 3 lines go by fast
https://github.com/caddyserver/caddy/blob/569ecdbd02a3bdf8589f785cca022240973dea4d/replacer_fuzz.go#L20-L22

But this one
https://github.com/caddyserver/caddy/blob/569ecdbd02a3bdf8589f785cca022240973dea4d/replacer_fuzz.go#L23
is the slow one

**Comment by mholt** [comment]
Interesting, so chopping the data in half... possibly unclosed placeholder or something?

**Comment by Mohammed90** [comment]
> Interesting, so chopping the data in half... possibly unclosed placeholder or something?

I wish there's a profiler that shows the number of times each line was hit to know the hot spots. The size of the input makes tracing the test case with debugger unreasonable.

**Comment by mholt** [comment]
I finally put this test case verbatim into a unit test (yes it's a ridiculously long line -- and I've replaced the NUL bytes with \x00 so the test will compile). I think the problem is that every empty placeholder -- of which there are a _million or so_ -- is being replaced with half the input -- which is also about a million bytes (give or take half of that, you get the idea) -- the allocations take time and it uses a ton of memory.

We could easily introduce artificial limits, but I think the notion of ~doubling the input size for every _empty placeholder_ and then having _millions of empty placeholders_ is not a real-world scenario. Nor is the use of `ReplaceAll()` with an untrusted input (or especially an untrusted "empty" value) a practical, real-world scenario.

In other words, I think this test case:

```
ReplaceAll(string(data[:len(data)/2]), string(data[len(data)/2:])) 
```

is not really representative of anything meaningful. Yes, it could be considered a bug technically, but the only good fix I can think of is "don't grow the string by orders of magnitude for every 2 bytes of input" :man_shrugging: 

Thanks for the fuzz testing! :+1:

**Comment by mholt** [comment]
Ok, so this is the same as #4168 -- it's just that (I think?) all the closing `}` are escaped (`\}`), so the search for the closing `}` goes to the end of the string (which is quite long), and it presumably never finds it, so then it iterates the rest of the string. Since the string is almost all `{ /}` characters, this is highly computationally complex.

This test case yields high CPU usage but less memory usage than #4168 -- because instead of building a monstrous string, it simply iterates the tens of thousands of bytes, tens of thousands of times.

Similar to #4168, I'm not sure how practical this is since placeholder input is generally trusted. This is technically a bug, I guess, but it's a GIGO bug. Maybe the only good fix is to terminate after N unclosed placeholders.

Hmm. Yeah, I think I will try that, why not. Fixed in 2dc747c
